### PR TITLE
[#1185] feat(server-common): Server supports Kerberos authentication

### PR DESCRIPTION
**Pull request by qqqttt123**
### What changes were proposed in this pull request?
Server supports Kerberos authentication. I referred to the implementation of Hadoop.
I will submit another pr about the client.

### Why are the changes needed?

Fix: #1185.

### Does this PR introduce _any_ user-facing change?
Added the document.

### How was this patch tested?
Add a new UT.

### CODE REVIEW

**Comment by mchades** [comment]
`openapi.yaml` should also be modified

**Comment by qqqttt123** [comment]
> `openapi.yaml` should also be modified

Openapi doesn't support this kind of authentication.

**Comment by qqqttt123** [comment]
@mchades @jerryshao @yuqi1129 @diqiu50 @FANNG1 @Clearvive Could you help me review this pr?

File: server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+      String authenticateToken = null;
+      if (serverToken != null && serverToken.length > 0) {
+        authenticateToken = Base64.getEncoder().encodeToString(serverToken);
+      }
+      if (!gssContext.isEstablished()) {
+        LOG.trace("SPNEGO in progress");
+        String challenge = AuthConstants.AUTHORIZATION_NEGOTIATE_HEADER + authenticateToken;
+        throw new UnauthorizedException("GssContext isn't established", challenge);
+      }
+      // Usually principal names are in the form 'user/instance@REALM' or 'user@REALM'.
```
Comment (qqqttt123) [inline server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java:137]: I simplify the logic of convert from principal to user.

File: server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+public class KerberosAuthenticator implements Authenticator {
+
+  public static final Logger LOG = LoggerFactory.getLogger(KerberosAuthenticator.class);
+  private final Subject serverSubject = new Subject();
+  private GSSManager gssManager;
+
+  @Override
+  public void initialize(Config config) throws RuntimeException {
+    try {
+      String principal = config.get(KerberosConfig.PRINCIPAL);
```
Comment (qqqttt123) [inline server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java:51]: We don't support `*` principal.

File: server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      String keytab = config.get(KerberosConfig.KEYTAB);
+      File keytabFile = new File(keytab);
+      if (!keytabFile.exists()) {
+        throw new IllegalArgumentException("Keytab doesn't exist: " + keytab);
+      }
+
```
Comment (qqqttt123) [inline server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java:57]: We remove the validation of rule names.

File: build.gradle.kts [hunk 4]
```diff
@@ -78,7 +78,8 @@ project.extra["extraJvmArgs"] = if (extra["jdkVersion"] in listOf("8", "11")) {
     "--add-opens", "java.base/sun.nio.ch=ALL-UNNAMED",
     "--add-opens", "java.base/sun.nio.cs=ALL-UNNAMED",
     "--add-opens", "java.base/sun.security.action=ALL-UNNAMED",
-    "--add-opens", "java.base/sun.util.calendar=ALL-UNNAMED"
+    "--add-opens", "java.base/sun.util.calendar=ALL-UNNAMED",
+    "--add-opens", "java.security.jgss/sun.security.krb5=ALL-UNNAMED"
```
Comment (jerryshao) [inline build.gradle.kts:82]: I think you should also update the start script under "bin".

File: common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+import javax.security.auth.login.AppConfigurationEntry;
+import javax.security.auth.login.Configuration;
+import javax.security.auth.login.LoginContext;
+import org.ietf.jgss.GSSException;
+import org.ietf.jgss.Oid;
+
+// Referred from Apache Hadoop KerberosTestUtils.java
+// hadoop-common-project/hadoop-auth/src/test/java/org/apache/hadoop/security/\
+// authentication/KerberosTestUtils.java
+public class KerberosUtils {
```
Comment (jerryshao) [inline common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java:36]: Still the same thing here, please point out which part did you modify for what purpose.

File: common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  public static <T> T doAs(String principal, String keyTabFile, final Callable<T> callable)
+      throws Exception {
+    LoginContext loginContext = null;
+    try {
+      Set<Principal> principals = new HashSet<>();
+      principals.add(new KerberosPrincipal(principal));
+      Subject subject = new Subject(false, principals, new HashSet<>(), new HashSet<>());
+      loginContext =
+          new LoginContext("", subject, null, new KerberosConfiguration(principal, keyTabFile));
+      loginContext.login();
```
Comment (jerryshao) [inline common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java:64]: Do we have to login everytime, is it a typical implementation?

File: server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+  private final Subject serverSubject = new Subject();
+  private GSSManager gssManager;
+
+  @Override
+  public void initialize(Config config) throws RuntimeException {
+    try {
+      String principal = config.get(KerberosConfig.PRINCIPAL);
+      String keytab = config.get(KerberosConfig.KEYTAB);
+      File keytabFile = new File(keytab);
+      if (!keytabFile.exists()) {
```
Comment (jerryshao) [inline server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java:54]: You should also check the permission file.

File: server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+      throw new UnauthorizedException("Blank token found", AuthConstants.NEGOTIATE);
+    }
+    try {
+      String serverPrincipal = KerberosUtil.getTokenServerName(clientToken);
+      return Subject.doAs(
+          serverSubject,
+          new PrivilegedExceptionAction<Principal>() {
+            @Override
+            public Principal run() throws Exception {
+              return validClientToken(serverPrincipal, clientToken);
```
Comment (jerryshao) [inline server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java:105]: `validateClientToken`?

File: server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosUtil.java [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+import javax.security.auth.kerberos.KerberosTicket;
+import javax.security.auth.kerberos.KeyTab;
+import org.apache.kerby.kerberos.kerb.keytab.Keytab;
+import org.apache.kerby.kerberos.kerb.type.base.PrincipalName;
+import org.ietf.jgss.GSSException;
+import org.ietf.jgss.Oid;
+
+// Referred from Apache Hadoop KerberosUtil.java
+// hadoop-common-project/hadoop-auth/src/main/java/org/apache/hadoop/\
+// security/authentication/util/KerberosUtil.java
```
Comment (jerryshao) [inline server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosUtil.java:45]: Also here, please point out which part did you modify for what purpose?

File: common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java (hunk 6 above)
Comment (qqqttt123) [reply common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java]: Split this into two methods.

File: common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+  private KerberosUtils() {}
+
+  public static final Oid GSS_SPNEGO_MECH_OID = getNumericOidInstance("1.3.6.1.5.5.2");
+  public static final Oid GSS_KRB5_MECH_OID = getNumericOidInstance("1.2.840.113554.1.2.2");
+  public static final Oid NT_GSS_KRB5_PRINCIPAL_OID =
+      getNumericOidInstance("1.2.840.113554.1.2.2.1");
+
+  // numeric oids will never generate a GSSException for a malformed oid.
+  // use to initialize statics.
```
Comment (jerryshao) [inline common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java:50]: Capitalize the first letter.

File: common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+  private static Oid getNumericOidInstance(String oidName) {
+    try {
+      return new Oid(oidName);
+    } catch (GSSException ex) {
+      throw new IllegalArgumentException(ex);
+    }
+  }
+
+  public static LoginContext login(String principal, String keyTabFile) throws LoginException {
```
Comment (jerryshao) [inline common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java:59]: Do you need a logout method?

File: server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+            public Principal run() throws Exception {
+              return validateClientToken(serverPrincipal, clientToken);
+            }
+          });
+    } catch (Exception e) {
+      throw new UnauthorizedException("Fail to validate the token", AuthConstants.NEGOTIATE);
+    }
+  }
+
+  private Principal validateClientToken(String serverPrincipal, byte[] clientToken)
```
Comment (jerryshao) [inline server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java:124]: This method doesn't only validate the token, but also return the user principal if the check is passed. I would suggest use a better method name.

File: common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java (hunk 11 above)
Comment (qqqttt123) [reply common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java:59]: I will call `LoginContext.logout`

File: server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java (hunk 12 above)
Comment (qqqttt123) [reply server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java:124]: How about `retrievePrincipalFromToken`?

File: common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java (hunk 10 above)
Comment (qqqttt123) [reply common/src/main/java/com/datastrato/gravitino/auth/KerberosUtils.java:50]: Fixed.

**Review by jerryshao — APPROVED** [review]

**Pull request merged**